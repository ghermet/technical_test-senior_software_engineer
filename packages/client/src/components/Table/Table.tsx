import React from 'react'
import { ctwMerge } from '../../utility/ctw-merge'

export const Table = ({
  children,
  ...props
}: React.DetailedHTMLProps<React.TableHTMLAttributes<HTMLTableElement>, HTMLTableElement> & {
  children: React.ReactNode | React.ReactNode[]
}) => {
  return <table {...props}>{children}</table>
}

Table.Headers = ({
  className,
  children,
  ...props
}: React.DetailedHTMLProps<React.HTMLAttributes<HTMLTableSectionElement>, HTMLTableSectionElement> & {
  children: React.ReactNode | React.ReactNode[]
}) => {
  return (
    <thead {...props} className={ctwMerge('block w-full bg-neutral4 laptop:rounded-lg', className)}>
      {children}
    </thead>
  )
}

Table.Header = ({
  className,
  children,
  ...props
}: React.DetailedHTMLProps<React.HTMLAttributes<HTMLTableHeaderCellElement>, HTMLTableHeaderCellElement> & {
  children: React.ReactNode | React.ReactNode[]
}) => {
  return (
    <th {...props} className={ctwMerge('body-bold-16 flex justify-start text-neutral2', className)}>
      {children}
    </th>
  )
}

Table.Body = ({
  className,
  children,
  ...props
}: React.DetailedHTMLProps<React.HTMLAttributes<HTMLTableSectionElement>, HTMLTableSectionElement> & {
  children: React.ReactNode | React.ReactNode[]
}) => {
  return (
    <tbody {...props} className={ctwMerge('block w-full', className)}>
      {children}
    </tbody>
  )
}

Table.Row = ({
  className,
  children,
  ...props
}: React.DetailedHTMLProps<React.HTMLAttributes<HTMLTableRowElement>, HTMLTableRowElement> & {
  children: React.ReactNode | React.ReactNode[]
}) => {
  return (
    <tr
      {...props}
      className={ctwMerge(
        `grid grid-flow-row grid-rows-1 gap-6 justify-normal items-center laptop:rounded-lg py-[1.25rem] px-[1.5rem] bg-white`,
        className
      )}
    >
      {children}
    </tr>
  )
}

Table.Cell = ({
  className,
  children,
  ...props
}: React.DetailedHTMLProps<React.HTMLAttributes<HTMLTableCellElement>, HTMLTableCellElement> & {
  children: React.ReactNode | React.ReactNode[]
}) => {
  return (
    <td {...props} className={ctwMerge('body-16 flex justify-start', className)}>
      {children}
    </td>
  )
}

export default Table
