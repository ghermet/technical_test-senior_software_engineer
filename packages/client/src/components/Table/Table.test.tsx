import React from 'react'
import { render } from '@testing-library/react'
import Table from './Table'

describe('Table', () => {
  it('should render a table element with the correct default classnames and props', () => {
    const { container } = render(
      <Table className="w-full">
        <thead>
          <tr>
            <th>data header</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>data value</td>
          </tr>
        </tbody>
      </Table>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a table headers element with the correct default classnames and props', () => {
    const { container } = render(
      <table>
        <Table.Headers className="w-full">
          <tr>
            <th>data header</th>
            <th>data2 header</th>
            <th>data3 header</th>
          </tr>
        </Table.Headers>
      </table>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a table header element with the correct default classnames and props', () => {
    const { container } = render(
      <table>
        <thead>
          <tr>
            <Table.Header className="w-fit">data header</Table.Header>
          </tr>
        </thead>
      </table>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a table body element with the correct default classnames and props', () => {
    const { container } = render(
      <table>
        <Table.Body className="w-full">
          <tr>
            <td>data value</td>
          </tr>
        </Table.Body>
      </table>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a table row element with the correct default classnames and props', () => {
    const { container } = render(
      <table>
        <tbody>
          <Table.Row>
            <td>data value</td>
            <td>data2 value</td>
            <td>data3 value</td>
          </Table.Row>
        </tbody>
      </table>
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a table cell element with the correct default classnames and props', () => {
    const { container } = render(
      <table>
        <tbody>
          <tr>
            <Table.Cell>data value</Table.Cell>
          </tr>
        </tbody>
      </table>
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
