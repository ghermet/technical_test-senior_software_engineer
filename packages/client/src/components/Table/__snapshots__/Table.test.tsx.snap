// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render a table body element with the correct default classnames and props 1`] = `
<table>
  <tbody
    class="block w-full"
  >
    <tr>
      <td>
        data value
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table should render a table cell element with the correct default classnames and props 1`] = `
<table>
  <tbody>
    <tr>
      <td
        class="body-16 flex justify-start"
      >
        data value
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table should render a table element with the correct default classnames and props 1`] = `
<table
  class="w-full"
>
  <thead>
    <tr>
      <th>
        data header
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        data value
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table should render a table header element with the correct default classnames and props 1`] = `
<table>
  <thead>
    <tr>
      <th
        class="body-bold-16 flex justify-start text-neutral2 w-fit"
      >
        data header
      </th>
    </tr>
  </thead>
</table>
`;

exports[`Table should render a table headers element with the correct default classnames and props 1`] = `
<table>
  <thead
    class="block bg-neutral4 laptop:rounded-lg w-full"
  >
    <tr>
      <th>
        data header
      </th>
      <th>
        data2 header
      </th>
      <th>
        data3 header
      </th>
    </tr>
  </thead>
</table>
`;

exports[`Table should render a table row element with the correct default classnames and props 1`] = `
<table>
  <tbody>
    <tr
      class="grid grid-flow-row grid-rows-1 gap-6 justify-normal items-center laptop:rounded-lg py-[1.25rem] px-[1.5rem] bg-white"
    >
      <td>
        data value
      </td>
      <td>
        data2 value
      </td>
      <td>
        data3 value
      </td>
    </tr>
  </tbody>
</table>
`;
